### PR TITLE
Add colorbar annotation example plot to gallery

### PR DESCRIPTION
**Pull request by wsoofi**
## PR Summary

Add an example plot to gallery showing how to annotate a colorbar.  

While the `matplotlib.colorbar.Colorbar.add_lines` function cleanly allows the user to add annotations to both a contour plot and its associated colorbar, adding an annotation to the colorbar alone requires a few steps. This example plot illustrates how to annotate **just** the colorbar. Addresses issue #12462. 

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: examples/images_contours_and_fields/colorbar_annotation_demo.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+"""
+=========================
+Colorbar Annotation Demo
+=========================
```
Comment (dstansby) [inline examples/images_contours_and_fields/colorbar_annotation_demo.py]: It looks like this title underline/overline has one too many characters (it has to exactly match the title length)

File: examples/images_contours_and_fields/colorbar_annotation_demo.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# ------------
+#
+# References
+# """"""""""
+#
+# The use of the following functions, methods and classes is shown
+# in this example:
+
+import matplotlib
+matplotlib.colorbar
```
Comment (dstansby) [inline examples/images_contours_and_fields/colorbar_annotation_demo.py]: I think the file needs a newline at the end too for the docs to build.

File: examples/images_contours_and_fields/colorbar_annotation_demo.py (hunk 1 above)
Comment (wsoofi) [reply examples/images_contours_and_fields/colorbar_annotation_demo.py]: Oh, I didn't catch that! Thanks!

Comment (wsoofi) [reply examples/images_contours_and_fields/colorbar_annotation_demo.py]: I've updated this.

File: examples/images_contours_and_fields/colorbar_annotation_demo.py (hunk 2 above)
Comment (wsoofi) [reply examples/images_contours_and_fields/colorbar_annotation_demo.py]: I have updated this too.

**Comment by wsoofi** [comment]
Hi all, this is my first time working on this repo so please forgive the questions. Everything is passing except the `flake8` test in the Travis CI build.  I noticed that the failure is due to a module level import not being at the top of the file (because it's in the **References** section at the bottom of the file). But it looks like several images in the gallery (such as `contour_image.py`) follow this convention.  Is it standard practice to ignore this error? @dstansby

**Comment by timhoffm** [comment]
You should add an exception for E402 to the file .flake8.

**Comment by jklymak** [comment]
Just to clarify.  You can add an e402 exception for this file.  Don’t add a general one!

**Comment by wsoofi** [comment]
Got pytest failures (`test_stem[png-w/o line collection]`) which seem unrelated to the changes I made.  I am closing and re-opening this PR to re-trigger the tests, just in case these were spurious errors.

**Comment by wsoofi** [comment]
Hi everyone,
All checks are passing except the coverage check for the `tests` directory.  Do we typically need to add tests when adding example plots in the gallery? Does coverage computation include documentation directories?

File: examples/images_contours_and_fields/colorbar_annotation_demo.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+# Build filled contour plot
+surfplot = plt.contourf(X, Y, Z, cmap=plt.cm.binary_r)
+cbar = plt.colorbar(surfplot)
+
+# Find min and max values on colorbar
+cmin = np.float(cbar.ax.get_yticklabels()[0].get_text())
+cmax = np.float(cbar.ax.get_yticklabels()[-1].get_text())
+
+# Use min/max values to calculate annotation location
+loc = (desired_cbar_yvalue - cmin) / (cmax - cmin)
```
Comment (jklymak) [inline examples/images_contours_and_fields/colorbar_annotation_demo.py]: I agree that you need to do this for `contour` colorbars, but its not true for `pcolormesh` colorbars.  So there needs to be more explanation here.  And frankly, I think the correct thing is for contour colorbars to be rendered between vmin and vmax.  So I'm a little leery about adding an example that shows how to get around inconsistent behaviour.

File: examples/images_contours_and_fields/colorbar_annotation_demo.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+Colorbar Annotation Demo
+========================
+
+Demonstrates how one can annotate the colorbar without needing to also add
+lines to a contour plot.
+
+Note that this is not a typical use case. The more common use case for
+annotations is to add lines to both a contour plot *and* the corresponding
+location on the colorbar. In this more common use case,
+`~matplotlib.colorbar.Colorbar.add_lines` can be used.
```
Comment (jklymak) [inline examples/images_contours_and_fields/colorbar_annotation_demo.py]: `add_lines` requires a `ContourSet` so I don't understand this comment.  Either it need to be expanded, or removed.

**Comment by ImportanceOfBeingErnest** [comment]
This seems to be a real edgecase. The workaround shown here is a creative hack, but do we want to teach people to use the tick**labels** on an axis to get a propper scale? 

Is there no example to show how to use `add_lines`? If not, *that* is what should be added. If so, can it simply be appended by something like the below?

```
import matplotlib.pyplot as plt
import numpy as np

# Create sample values to plot
x = np.linspace(0, 1, 100)
y = np.linspace(0, 1, 100)
X, Y = np.meshgrid(x, y)
Z = np.exp(-X**2 - Y**2)

# Choose a value to annotate
desired_cbar_yvalue = 0.90

# Build filled contour plot
surfplot = plt.contourf(X, Y, Z, cmap=plt.cm.binary_r)
cbar = plt.colorbar(surfplot)

line = plt.contour(X, Y, Z, levels=[desired_cbar_yvalue], colors="red", linewidths=5)
cbar.add_lines(line)
line.collections[0].set_visible(False)

plt.show()
```

**Comment by jklymak** [comment]
Well, the fault is mine - this *used* to work on all colorbars, but then I changed the colorbars to actually be proper axes between vmin and vmax, instead of faked axes that were really rendered between 0 and 1.  However, I've not fixed the `ContourSet` colorbars to behave the same way.

**Comment by ImportanceOfBeingErnest** [comment]
Not sure I understand what you mean. So for normal colorbars you just do `cbar.ax.axhline(0.9)`. I understand that this is not possible for the contour colorbars. Hence my proposal above.

**Comment by jklymak** [comment]
Right but that only started to work for 3.0.

**Comment by wsoofi** [comment]
Thanks for the very timely and thorough comments! My solution was based on the suggestion from [this issue comment](https://github.com/matplotlib/matplotlib/issues/12462#issuecomment-428546529) that we could use an example in the gallery showing how to grab the colorbar axes and then add a line to it. I agree that the solution contains a workaround, but my impression was that it was a required workaround given the current behavior of `contour` colorbars.

I have a few questions about next steps.

1) @jklymak, with your last comment above, are you saying that @ImportanceOfBeingErnest’s solution is not backward compatible with versions of matplotlib prior to 3.0 and therefore should not be included in the gallery? 

2) To answer @ImportanceOfBeingErnest's question, there _is_ an add_lines demo [here](https://matplotlib.org/gallery/images_contours_and_fields/contourf_demo.html).  Should the example in this PR be appended there instead of standing on its own?

3) It looks like @ImportanceOfBeingErnest’s solution involves building two contour plots, one on top of the other, and then hiding part of the top contour plot.  Is this layering method considered idiomatic matplotlib? If so, it makes sense to me to go with that rather than the one in this PR (assuming we’d be okay with including this solution, given question # 1 above).  If not, then which method do we think would be better to include as an exemplar, given that both seem to work?

Thanks again for all the helpful comments!

File: examples/images_contours_and_fields/colorbar_annotation_demo.py (hunk 3 above)
Comment (wsoofi) [reply examples/images_contours_and_fields/colorbar_annotation_demo.py]: Thanks! I can certainly add more explanation here (depending perhaps on what we decide to do with this example, given the discussion below).

**Pull request closed without merging**